tomcat/src/main/java/camp/nextstep/http/domain/RequestLine.java

```diff
@@ -0,0 +1,133 @@
+    private static final int HTTP_PATH_INDEX = 1;
+    private static final int HTTP_PROTOCOL_ARG_INDEX = 2;
+    private static final int PROTOCOl_INDEX = 0;
+    private static final int HTTP_VERSION_INDEX = 1;
+    private static final int PATH_INDEX = 0;
+    private static final int QUERY_PARAM_INDEX = 1;
+
+    private HttpMethod method;
+    private String path;
+    private Protocol protocol;
+    private HttpVersion version;
+    private Map<String, String> queryParams;
+
+    private RequestLine(
+            HttpMethod method,
+            String path,
+            Protocol protocol,
+            HttpVersion version
+    ) {
+        this.method = method;
+        this.path = path;
+        this.protocol = protocol;
+        this.version = version;
+    }
+
+    private RequestLine(
+            HttpMethod method,
+            String path,
+            Protocol protocol,
+            HttpVersion version,
+            Map<String, String> queryParams
+    ) {
+        this.method = method;
+        this.path = path;
+        this.protocol = protocol;
+        this.version = version;
+        this.queryParams = queryParams;
+    }
+
+    public Map<String, String> getQueryParams() {
+        return queryParams;
+    }
+
+    public static RequestLine createRequestLineByRequestLineStr(String requestLineStr) {
+        String[] requestArgs = requestLineStr.split("\\s+");
+        if (requestArgs.length != REQUEST_LINE_ARGS_COUNT) {
+            throw new InvalidHttpRequestSpecException("http 스펙의 파라미터 개수가 유효하지 않습니다");
+        }
+
+        HttpMethod method = getHttpMethodFromStr(requestArgs[HTTP_METHOD_INDEX]);
+
+        String[] httpArgs = requestArgs[HTTP_PROTOCOL_ARG_INDEX].split("/");
+        if (httpArgs.length != HTTP_PROTOCOL_ARGS_COUNT) {
+            throw new InvalidHttpRequestSpecException("유효한 프로토콜 형식이 아닙니다");
+        }
+
+        Protocol protocol = getProtocolFromStr(httpArgs[PROTOCOl_INDEX]);
+        HttpVersion version = getHttpVersionFromStr(httpArgs[HTTP_VERSION_INDEX]);
+
+        String[] pathStrs = requestArgs[HTTP_PATH_INDEX].split("\\?");
+        if (pathStrs.length == 2) {
+            Map<String, String> queryParams = getQueryParams(pathStrs[QUERY_PARAM_INDEX]);
+
+            return new RequestLine(
+                    method,
+                    pathStrs[PATH_INDEX],
+                    protocol,
+                    version,
+                    queryParams
+            );
+        } else {
+            return new RequestLine(
+                    method,
+                    pathStrs[PATH_INDEX],
+                    protocol,
+                    version
+            );
+        }
+    }
+
+    private static HttpMethod getHttpMethodFromStr(String httpMethodStr) {
+        HttpMethod method = searchEnum(HttpMethod.class, httpMethodStr);
+        if (method == null) {
+            throw new InvalidHttpRequestSpecException("지원하는 메소드가 아닙니다");
+        }
+        return method;
```

Comment:
반환타입이 nullable하게 될 경우 매번 null을 신경써서 NPE를 방지해줘야하는데, non-nullable하게 설계해볼순 없을까요? 

```suggestion
        HttpMethod method = searchEnum(HttpMethod.class, httpMethodStr);
        if (method == HttpMethod.NONE) {
            throw new InvalidHttpRequestSpecException("지원하는 메소드가 아닙니다");
        }
        return method;
```

이런식으로 말이죠 ㅎㅎ